### DISCUSSION TITLE
unificontrol and PyScript

### DISCUSSION THREAD

**Discussion by nickolas1969** [custom-components/pyscript#526] [Q&A]
My first attempt with PyScript is to use the unificontrol module to make a script that logs into my home unifi controller and blocks/unblocks a MAC address passed to the script.

I added “unificontrol” to requirements.txt as stated in the docs for adding external modules.

in my script I followed the “hello world” service example: adding my import at the top of my unifi.py script, and altering the parameters to keep “action” and look for “block” or “unblock”, and the “id” changed to “mac” and would be the MAC address to act on.

the script works in my external Ubuntu server, so code is good. Does it sound like I did the external library import correctly?

**Comment by ALERTua** [comment]
To better analyze the cause please provide logs of the executed pyscript script, the script body, the requirements.txt body, and the pyscript directory tree screenshot. If you don't know how to provide any of these - please let me know, and I'll describe them in detail.

**Reply by nickolas1969** [reply]
This error originated from a custom integration.

Ahhh!! Ok, so be making a call to a rest API of another system, PyScript wants that done a different way. I need to find docs on how they want  "await hass.async_add_executor_job()`"?

Logger: custom_components.pyscript.file.unifi.block_mac
Source: custom_components/pyscript/eval.py:493
Integration: Pyscript Python scripting (documentation, issues)
First occurred: September 12, 2023 at 7:58:18 PM (3 occurrences)
Last logged: 9:40:50 AM

Exception in <file.unifi.block_mac> line 8: client.block_client(mac) ^ RuntimeError: Blocking calls must be done in the executor or a separate thread; Use `await hass.async_add_executor_job()`; at custom_components/pyscript/eval.py, line 1941: return func(*args, **kwargs)

**Reply by nickolas1969** [reply]
Here is the full script I am using.
```
import unificontrol

@service
def block_mac(action=None, mac=None):
    client = unificontrol.UnifiClient(host="10.0.0.5",username='<snip>', password='<snip>')
    log.info(f"block_mac: got action {action} mac {mac}")
    if action == "block" and mac is not None:
        client.block_client(mac)
    elif action == "unblock" and mac is not None:
        client.unblock_client(mac)
    client.logout()
```

**Reply by nickolas1969** [reply]
I did try wrapping the client.block_client() in the await hass.async_add_executor_job() in place, but that threw an error saying the call needed to be somewhere else. 

Maybe PyScript in HA isn't what I need to allow HA to reach outside it's world to change other API's, due to the async nature of how it's running.

**Reply by ALERTua** [reply]
You need this documentation block: https://hacs-pyscript.readthedocs.io/en/stable/reference.html#avoiding-event-loop-i-o
Similar answered question: https://github.com/custom-components/pyscript/discussions/419

**Reply by ralle12345** [reply]
@nickolas1969  Why don't you just use the unifi integration? It provides a switch entity for each and every wifi device. Using these you can just use the switch.turn_off service to block a client in your wifi.

**Comment by nickolas1969** [comment]
I did not know it existed…Nick Ellson - from iPhone (forgive typos)CCIE #20018; VCIX-NV, CNSENetwork Hobbyist"Educating Layer 8, one user at a time."On Sep 19, 2023, at 10:49 PM, ralle12345 ***@***.***> wrote:﻿
@nickolas1969  Why don't you just use the unifi integration? It provides a switch entity for each and every wifi device. Using these you can just use the switch.turn_off service to block a client in your wifi.
